text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,379 @@
+// Copyright (C) 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include <openvino/openvino.hpp>
+#include <cmath>
+#include <random>
+
+constexpr size_t BATCH_SIZE = 1;
+
+namespace {
+std::pair<ov::Tensor, ov::Tensor> tokenize(ov::InferRequest& tokenizer, std::string&& prompt) {
+    constexpr size_t BATCH_SIZE = 1;
+    tokenizer.set_input_tensor(ov::Tensor{ov::element::string, {BATCH_SIZE}, &prompt});
+    tokenizer.infer();
+    return {tokenizer.get_tensor("input_ids"), tokenizer.get_tensor("attention_mask")};
+}
+
+std::string detokenize(ov::InferRequest& detokenizer, std::vector<int64_t>& tokens) {
+    constexpr size_t BATCH_SIZE = 1;
+    detokenizer.set_input_tensor(ov::Tensor{ov::element::i64, {BATCH_SIZE, tokens.size()}, tokens.data()});
+    detokenizer.infer();
+    return detokenizer.get_output_tensor().data<std::string>()[0];
+}
+
+std::vector<float> softmax(const ov::Tensor& logits, float temperature) {
+    float* logits_data = logits.data<float>();
+    int size = logits.get_size();
+    
+    double sum_exp = 0.0;
+    for (int i = 0; i < size; i++) {
+        sum_exp += std::exp(logits_data[i] / temperature);
+    }
+    
+    std::vector<float> probabilities;
+    for (int i = 0; i < size; i++) {
+        double probability = exp(logits_data[i] / temperature) / sum_exp;
+        probabilities.push_back(probability);
+    }
+    return probabilities;
+}
+
+int random_sample(const ov::Tensor& logits, float temperature) {
```

Comment:
is it going to be used as next improvement?

Reply:
it will not be used, i will remove it

Reply:
removed